### ISSUE TITLE
Some "reach the end" errors are not reported. Fix it, or twist the spec?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#4091]
Here is an issue that I could have created in the SDK repository. I'm creating this issue instead because it would be a breaking change to fix the implemented behavior such that it is aligned with the specified behavior. 

So, @dart-lang/language-team, should we go ahead and change the implementations to follow the spec? Or should we change the spec to match one of the specifications? (They don't agree, so we can't just specify the existing behavior.)

-----
[The following is a potential SDK issue, with labe 'area-meta' and subissues for the analyzer and for the CFE]

Consider the following library:

```dart
import 'dart:async';

void f1() {}
dynamic f2() {}
Null f3() {}

int f4() {} // Error ++
Object? f5() {} // Error -*
FutureOr<Object>? f6() {} // Error -*
FutureOr<Object?> f7() {} // Error -*
FutureOr<void> f8() {} // Error --

class A {
  factory A.f4() {} // Error ++
}

Future<void> g01() async {}
Future<dynamic> g02() async {}
Future<Null> g03() async {}

FutureOr<void> g04() async {}
FutureOr<dynamic> g05() async {}
FutureOr<Null> g06() async {}

Future<void>? g07() async {}
Future<dynamic>? g08() async {}
Future<Null>? g09() async {}

FutureOr<void>? g10() async {}
FutureOr<dynamic>? g11() async {}
FutureOr<Null>? g12() async {}

void g13() async {}
dynamic g14() async {}

Future<int> g17() async {} // Error ++
Future<Object?> g18() async {} // Error -*
Future<FutureOr<Object>?> g19() async {} // Error -*
Future<FutureOr<Object?>> g20() async {} // Error -*
Future<FutureOr<void>> g21() async {} // Error --

FutureOr<int> g23() async {} // Error ++
FutureOr<Object?> g24() async {} // Error -*
FutureOr<FutureOr<Object>?> g25() async {} // Error -*
FutureOr<FutureOr<Object?>> g26() async {} // Error -*
FutureOr<FutureOr<void>> g27() async {} // Error --

Future<int>? g29() async {} // Error -+
Future<Object?>? g30() async {} // Error -*
Future<FutureOr<Object>?>? g31() async {} // Error -*
Future<FutureOr<Object?>>? g32() async {} // Error -*
Future<FutureOr<void>>? g33() async {} // Error --

FutureOr<int>? g35() async {} // Error -+
FutureOr<Object?>? g36() async {} // Error -*
FutureOr<FutureOr<Object>?>? g37() async {} // Error -*
FutureOr<FutureOr<Object?>>? g38() async {} // Error -*
FutureOr<FutureOr<void>>? g39() async {} // Error --
```

Each line marked with 'Error' [should be reported as a compile-time error](https://github.com/dart-lang/language/blob/9558905ec2151a2b6b1c5154a44db5a2deb00388/specification/dartLangSpec.tex#L1996). However, the CFE does not report the lines where the comment has a '-' immediately after `Error `, and the analyzer does not report the lines where `Error ` is followed by a `-` one position later (as in `Error x-` where `x` can be anything). Finally, the analyzer reports a warning rather than an error in lines where `Error` is followed by `*` in the same position (as in `Error x*`).

The underlying reason is that it is an error to "return nothing" when the return type isn't `void`, `dynamic`, or `Null` (for a non-`async` non-generator function), or when the future value type of the function isn't `void`, `dynamic`, or `Null` (for an `async` non-generator function).

This error does not depend on the soundness of the implied semantics (which is to return null), it is intended to give developers a heads-up even in a function whose return type is, say, `Object?`. In such a function it's perfectly sound to return null, but the assumption is that the end of the function has been reached because of a bug, not because the developer intended to return null. A remedy could be to add an explicit `return null;` at the end of the body which would eliminate the error, even though it doesn't change the behavior at run time at all.

**Comment by stereotype441** [comment]
I would be open to the possibility of fixing the implementations to match the spec, provided that (a) we do the fix in a language-versioned way, so that users have control of when they have to migrate their code, and (b) if we think that a lot of code is going to be affected, we provide some tooling to help users with the transition.

Perhaps a reasonable first step would be to do a trial implementation of the correct rules that reports a warning rather than an error, and then use find_violations in google3 to measure how much code will be affected in practice.

**Comment by lrhn** [comment]
I'd want to make sure our specification is exactly what we want, before we try to make changes to behavior. (It's probably is.)

As I understand it:
* The spec treats `return;` and "reaching-body-end" the same. All errors here are places where the implementations don't. (If I change the `{}` body to `{return;}`, every marked case becomes an error in both CFE and analyzer.) So we should fix this for the consistency. 👍 
* You can write `return;` if the returned value type (the return type of a sync function, the future-value-type of an `async` function) is `void`, `dynamic` or `Null`. (Is that "is" a type-equals or **Norm**-equivalence? I think it's equals, otherwise a return-type of `FutureOr<void>` would count.)
* The errors here are all allowing reaching the end whenever the return type is *nullable*, not just when it's `Null`, treating it as an implicit `return null;`, not an implicit `return;`. Plus the analyzer giving warnings when that nullable type is not `void`, and seemingly doing **Norm** normalization on the return type (since `FutureOr<void>` avoids the warning.)

Using type-equality is always dangerous, because our static type system implementations are not very good at it. Or rather, it's not very *spec-compliant* at it, because it does some normalization that's really part of **Norm**, and not other normalization. I was going to say "not very consistent", but it is actually pretty consistent.

The types `Never?` and `Null` are *not* equal. They are **Norm** equivalent. Our front-ends don't distinguish them.
And they may consider some `Null`- or `Never`-bounded types as the same thing.

More tests, not showing much new.
```dart
void gen<N extends Null, V extends Never>() {
  Null g40() {}
  Never? g41() {} // Error --  
  N g50() {} // Error ++        
  V? g51() {} // Error --         
  N? g52() {} // Error --

  Null g40r() {return;}
  Never? g41r() {return;} // Error --
  N g50r() {return;} // Error ++
  V? g51r() {return;} // Error ++
  N? g52r() {return;} // Error ++

  Future<Null> g60() async {}
  Future<Never?> g61() async {} // Error --
  Future<N> g70() async {} // Error ++
  Future<V?> g71() async {} // Error --
  Future<N?> g72() async {} // Error --

  Future<Null> g60r() async {return;}
  Future<Never?> g61r() async {return;} // Error --
  Future<N> g70r() async {return;} // Error ++
  Future<V?> g71r() async {return;} // Error ++
  Future<N?> g72r() async {return;} // Error ++
}
```
The `g50`/`g70` errors are good, the functions aren't sound.

The `g41r`/`g61r` missing errors are probably due to front ends considering `Never?` *the same* as `Null`.

For the rest, it's just "allow reaching end if return type nullable" plus analyzer doing **Norm** before checking if `== Null` (so no warnings).

(I also have opinions about not being allowed `return;` for a type of `FutureOr<void>`, but that's a bigger change.)

**Comment by jakemac53** [comment]
> (I also have opinions about not being allowed `return;` for a type of `FutureOr<void>`, but that's a bigger change.)

Yeah this is the only one that seems weird to me, it seems wrong to be forced to explicitly return a value (even if it is `null`) from a `void` (or `FutureOr<void>`, or `void|int` if we had full union types) function. I don't think most users think of `void` as being a real value, null or otherwise.

**Comment by munificent** [comment]
**TL;DR: I find missing return errors to be very common mistakes in my code, so I want the language to be strict. If a function *can* return a usable value, it should yell at me unless I do. Basically, unless the return type is `void` or something wrapped around `void`, it should probably be an error to not have a return value.**

I found your encoding a little opaque. :) I tried expanding it. Let me know if I got it wrong. Assuming I have it right, here's my opinion:

```dart
// I think these should all be errors:
int f4() {} // CFE error, analyzer error.
Object? f5() {} // CFE silent, analyzer warning.
FutureOr<Object>? f6() {} // CFE silent, analyzer warning.
FutureOr<Object?> f7() {} // CFE silent, analyzer warning.

// I'm not sure about this one. Lean towards no error:
FutureOr<void> f8() {} // CFE and analyzer silent.

class A {
  // Error:
  factory A.f4() {} // CFE error, analyzer error.
}

// No error:
Future<void> g01() async {}

// Errors:
Future<dynamic> g02() async {}
Future<Null> g03() async {}

// No error:
FutureOr<void> g04() async {}

// Errors:
FutureOr<dynamic> g05() async {}
FutureOr<Null> g06() async {}

// No error. But also really weird:
// "Eventually, I may return nothing... or not" is a deeply odd contract. :)
Future<void>? g07() async {}

// Errors:
Future<dynamic>? g08() async {}
Future<Null>? g09() async {}

// Uh... flip a coin? "I may return nothing right now, or later, or not." WAT:
FutureOr<void>? g10() async {}

// Errors:
FutureOr<dynamic>? g11() async {}
FutureOr<Null>? g12() async {}

// No error:
void g13() async {}

// Error:
dynamic g14() async {}

// Errors:
Future<int> g17() async {} // CFE error, analyzer error.
Future<Object?> g18() async {} // CFE silent, analyzer warning.

// LOLWUT:
Future<FutureOr<Object>?> g19() async {} // CFE silent, analyzer warning.
Future<FutureOr<Object?>> g20() async {} // CFE silent, analyzer warning.
Future<FutureOr<void>> g21() async {} // CFE and analyzer silent.

// Errors:
FutureOr<int> g23() async {} // CFE error, analyzer error.
FutureOr<Object?> g24() async {} // CFE silent, analyzer warning.
FutureOr<FutureOr<Object>?> g25() async {} // CFE silent, analyzer warning.
FutureOr<FutureOr<Object?>> g26() async {} // CFE silent, analyzer warning.

// No error:
FutureOr<FutureOr<void>> g27() async {} // CFE and analyzer silent.

// Errors:
Future<int>? g29() async {} // CFE silent, analyzer error.
Future<Object?>? g30() async {} // CFE silent, analyzer warning.
Future<FutureOr<Object>?>? g31() async {} // CFE silent, analyzer warning.
Future<FutureOr<Object?>>? g32() async {} // CFE silent, analyzer warning.

// No error:
Future<FutureOr<void>>? g33() async {} // CFE and analyzer silent.

// Errors:
FutureOr<int>? g35() async {} // CFE silent, analyzer error.
FutureOr<Object?>? g36() async {} // CFE silent, analyzer warning.
FutureOr<FutureOr<Object>?>? g37() async {} // CFE silent, analyzer warning.
FutureOr<FutureOr<Object?>>? g38() async {} // CFE silent, analyzer warning.

// Error:
FutureOr<FutureOr<void>>? g39() async {} // CFE and analyzer silent.
```

**Comment by lrhn** [comment]
> **Basically, unless the return type is `void` or something wrapped around `void`, it should probably be an error to not have a return value.**

Yes please! Just `void` or a union type where `void` is one of the parts. 
Not even `X?` where `X` is a type variable bounded by `void`. (Also, don't use `void` as a bound!)

More breaking, especially when it doesn't allow `dynamic`.

**Comment by leafpetersen** [comment]
Focusing first on just the inconsistencies: these are the only places where we currently have inconsistent implementations:
```dart
Future<int>? g29() async {} // Error -+
//           ^^^
// [analyzer] COMPILE_TIME_ERROR.BODY_MIGHT_COMPLETE_NORMALLY

FutureOr<int>? g35() async {} // Error -+
//             ^^^
// [analyzer] COMPILE_TIME_ERROR.BODY_MIGHT_COMPLETE_NORMALLY
```
These cases will always result in runtime errors, and already have an analyzer error, so fixing the lack of an error for these in the CFE should be an easy change to roll out.

Focusing secondly on the behavior, I was curious about how these tests behaved in three situations: first where the body was `{}`, secondly where it was `{return;}` and thirdly where the body was `{return null;}`.  Other than the two inconsistent cases above (where the CFE treats the empty body and return of null cases differently) both implementations uniformly emit errors for `{}` iff they emit errors for `{return null;}`.  That is, they seem de facto to be using "return an error for reaching the end of the body if `return null;` would be an error" rather than "if `return;` would be an error" as specified.  I don't know if there was some change in the spec along the way that led to this or if it's just coincidence?

How do people feel about `{}` vs `{return;}` vs `{return null;}?  Semantically, there's no difference.  Do we feel that they should all have the same errors?  Or do we feel there is a moral difference?

I made a test file from the original tests from @eernstg plus the `return;` and `return null;` cases with the current analyzer and cfe behaviors documented [here](https://dart-review.googlesource.com/c/sdk/+/385361) for discussion.

**Comment by lrhn** [comment]
> How do people feel about `{}` vs `{return;}` vs `{return null;}`? Semantically, there's no difference. 

Runtime semantics are the same, ale return the value `null`.
Semantic specification differs, `return;` "returns without a value" and `return null;`  "returns the value `null`". It's just that the caller can't tell the difference.

I would have expected a difference for static semantics, where `return;` would infer a return type of `void`. In practice they all infer a return type of `Null`, so there really is no difference by themselves. 
For a body of `if (b) return 2;` fingers by one of `return;`, `return null;` or nothing, there _is_ a difference, where `return;` becomes an error, and the other two infer a return type of `int?`.


> Do we feel that they should all have the same errors? Or do we feel there is a moral difference?

Yes, there is a moral difference. 
They should _not_ have the same errors. 

The type `void` exists to document the intent of not returning a value from the function. It is the only type with that intent.
The `return;` and reaching the end without returning both represent not actually returning a value. 
Reaching the end accidentally where a value is intended to be returned should be treated as a very likely programming error.

All of this exist to help users not have errors. We could remove `void` today, making it an alias for `Object?` and `return;` a shorthand for `return null;`, and all valid programs should keep running.
Users would lose the expressibility and help they get from `void`.

We should either make `void` the best damn help out can be, or we should admit that we have made it less useful because we won't want to break existing code. 
(Remembering that the choices we made at Dart 2.0 to not break existing programs were made in a language without versioning. We didn't allow `return;` with a return type of `Null` because we wanted to, but because we couldn't get away with not allowing it. _Maybe_ we can today.)

There is the alternative of changing how we look at `void`, at what intent it implies, to really mean "something nullable", because `null` represents "no value".
Then having a branch reach the end of a function with a nullable return type is fine, that returns "no value" aka `null`.
An actual `void` is treated something like `X?` where `X` is unknown, the only value you know it can have is `null`.

**Comment by leafpetersen** [comment]
> Semantic specification differs, `return;` "returns without a value" and `return null;` "returns the value `null`". It's just that the caller can't tell the difference.

To be clear, the semantic specification *does not differ*.  As you yourself say below the quoted bit above, "return no value" is also known as "return null".  Two different words for the same thing do not make for a semantic difference.  A syntactic one perhaps.  :)

> returns "no value" aka `null`.

In terms of helping users, it seems to me that the biggest low hanging fruit here are the `async` functions with non-nullable future value type which have any one of the three variants, since actually returning `null` from those functions will *always* be a runtime error, right? 

I think I hear from @lrhn a desire to change how `return;` behaves, which in some sense is outside the scope of this right?  I believe that for better or worse, all of those are working as specified?  I'm open to considering something there though.

Setting aside that, what about the `{}` cases and `return null` cases?  Concretely: do we believe those two should always behave the same?  Or differently?  In either case, how should they behave?

**Comment by lrhn** [comment]
> To be clear, the semantic specification does not differ. 

I'm saying that the semantic definition does distinguish `return;` from `return null;`.
It may be a distinction without a semantic difference, but it exists for a reason.
There is an intentional difference even if there's no extensional difference in behavior. 

Even in Dart 1, where semantics ignored type annotations, it was an error for a function to contain both a `return;` and a `return expr;`, because the former was intended for functions whose body does not return a value, and the latter was intended for functions whose body returns a value, and each function body should be only one of those. 

Since all function _calls_ evaluate to a value, calling a function whose body does not return a value does evaluate to `null`.

Now, the _specification_ has not always been good at expressing that intentional difference, because it only cared about expressing the runtime semantics.

> I think I hear from @lrhn a desire to change how return; behaves, which in some sense is outside the scope of this right?

Correct. But the changes I want are mostly to _reject_ more programs, which means that I definitely do not want to _allow_ more programs than what the current specification allows, even if implementations currently do allow to much. Fix the implementations.
(If we are going to do breaking bug-fixes, we can consider whether that's a good time to do more changes too, but it is outside the door of this question.)

> Setting aside that, what about the `{}` cases and `return null` cases? Concretely: do we believe those two should always behave the same? Or differently? In either case, how should they behave?

Inside the scope of this question, I think `{}` and `return;` should always behave the same, and I think that the specification already days so.

There are cases in the current specification where we disallow `return;` and allow `return null`, when the return type is nullable but not `void`, `dynamic` or `null`.
We should keep that distinction, and make `{}` match `return;`.

Basically: fix the implementations to match the specification. 

I do not want to allow a `return;` or `{}` in a function with return type `int?`.

(And the bigger change I'd _want_ to do, outside the scope of this bug report, is to only allow `return;` and `{}` of the return value type is `void`, or a union type containing `void`. Might have to accept `dynamic` too as long as writing no return type gives you `dynamic`. And also disallowing `return expr;` if the return value type is `void`, no matter which type `expr` has. And optimally, have any expression with static type `void` or context type `void` discard it's value and evaluate to `null`, as a "void coercion".)

**Comment by leafpetersen** [comment]
> That's specified as part of function invocation, not a property of the function or its body. If execution of a (plain, aka. synchronous non-generator) function body completes without returning a value, then the invocation evaluates to `null`.

This is really a pointless argument, but for the record, this is entirely not an accurate description of the semantics (as implemented) of Dart programs.  Beyond the simple fact that this is not what actually happens, this is also quite observable in the case of the `async` functions.  It is not possible without making all calls reflective to accurately model the way that async functions behave if you model it as "the function returns no value, and at the call site no value gets turned into null", because the implicit cast which causes the return to fail depends on type information which is not available at the call site.

> Fix the implementations to make `{}` equivalent to `return;` in every way.

Note that `{}` is just one special case of programs that run off the end of the block.  I assume you don't mean to just special case `{}`?

So then in the following, would you suggest that f1 be rejected iff f2 is rejected?
```dart
SomeType f1(bool b) {
   if (b) return;
}

SomeType f2 (bool b) {
  if (b) return;
  return;
}
```

What then do we say about:
```dart
SomeType f3(bool b) {
  if (b) return something;
}
```

If we say that there is an implicit `return` at the end of every path that exits a function without returning, then `f3` becomes an error (it currently is accepted).  Is that what we want?   If so, we'll need to do some work to figure out how breaking this is.

**Comment by lrhn** [comment]
Agree that `async` functions really do return a `Future`, whether or not their body returns a value. That's defined in the rules for invoking an `async` function, which then executes the body, and if the body completes normally (reached the end) or completes by returning without a value, the future is completed with `null`, and if the body complete by returns with a value, the future is completed with that value. That's part of the semantics of invoking an `async` function, but not part of the semantics of the body code itself. (Which, again, is a distinction without a difference for actual users.)

>  I assume you don't mean to just special case `{}`?

Yes. It just got long-winded to keep saying "reaches-the-end-of-the-body". The `{}` is a representative of the case where control flow reaches the end of the body.

So about the following, where `SomeType` can be any type.
```dart
SomeType f1(bool b) {
   if (b) return;
}

SomeType f2 (bool b) {
  if (b) return;
  return;
}
```
I would indeed have `f1` rejected iff `f2` is rejected.
Not surprising since I have consistently said that reaching the end and having a `return;` should be treated equivalently.

And the `return;` would be an error if `SomeType` is not `void`, `dynamic` or `Null`, and so would reaching the end.
(Like, if I understand it correctly, it's currently specified.)

For:
```dart
SomeType f3(bool b) {
  if (b) return something;
}
```
that would be an error if `SomeType` is not `void`, `dynamic` or `Null`,
and it would, as usual, be an error if the type of `something` is not assignable to that type.

> then `f3` becomes an error (it currently is accepted). Is that what we want?

That's exactly what I want. I've never wanted anything else.

I also *want* to make `return something;` always be an error if `SomeType` is `void`. Currently it's only an error if the static type of `something` isn't `void`, `dynamic` or `Null`, but it's not good for readability to look like it's returning a value, when in fact it's a `void` function. But again that's not the scope for this issue.

For the question "Fix implementations or change spec?", my answer is "fix implementations". Allowing the current implementation isn't helping anyone avoid errors, and it's not allowing good, readable code.


The spec (without comments) for a function where the return type *T* and *S* is the static type of an expression `e`, is:
> ```latex
> \Case{Synchronous non-generator functions}
> Consider the case where $f$ is a synchronous non-generator function
> (\ref{functions}).
> It is a compile-time error if $s$ is \code{\RETURN;},
> unless $T$ is \VOID, \DYNAMIC, or \code{Null}.
>
> It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
> $T$ is \VOID,
> and $S$ is neither \VOID, \DYNAMIC, nor \code{Null}.
>
> It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
> $T$ is neither \VOID, \DYNAMIC, nor \code{Null},
> and $S$ is \VOID.
>
> It is a compile-time error if $s$ is \code{\RETURN{} $e$;},
> $S$ is not \VOID,
> and $S$ is not assignable to $T$.
> ```
and elsewhere
> ```latex
> Unless it is statically known that the body of the function
> cannot complete normally
> (\commentary{that is, it cannot reach the end and ``fall through''},
> cf.~\ref{statementCompletion}),
> it is a compile-time error if
> the addition of \code{\RETURN;} at the end of the body
> would be a compile-time error.
> ```

That's the current spec. I think we should implement that rather than modify the spec to match missing errors from the implementations.

(It shouldn't be specified like that, rather it should repeat the text from `return;` in the "reaches the end of the body" case, so it's a compile-time error if control flow can reach the end of a synchronous non-generator function, or cannot be proven not to reach it, which is as much as this spec can say without a specification of inference, and the return type is not one of `void`, `dynamic` or `Null`. But the effect is the same as writing a `return;` at the end.)
